### PR TITLE
feat: Add status info of OOM

### PR DESCRIPTION
**Pull request by youngjun0627**
### issue
https://github.com/lablup/backend.ai/issues/265

### idea
I obtain kernel's status info from docker-events-subscriber. Event-data from the subscriber contains 'Action' key, that means 'create' or 'die' or 'oom'. When Out-of-memory occurs in a kernel, 'Action' key contains value of 'oom'. if check that, I insert status info of 'oom' to kernel_object's termination_reason. And, update column of status_info from kernels table.

### result
I test my code using [BFS algorithm](https://www.acmicpc.net/problem/1726). I generate test-case, and two solution code, one of them is efficient in terms of memory, the other is not. Following images show difference results.

- efficient code : 
image

- not efficient code:
image

The progress bar in the images is related to https://github.com/lablup/backend.ai/issues/227

### CODE REVIEW

File: src/ai/backend/agent/docker/agent.py [hunk 1]
```diff
@@ -1228,6 +1228,9 @@ async def handle_action_die(kernel_id: KernelId, evdata: Mapping[str, Any]) -> N
                         kernel_id = await get_kernel_id_from_container(container_name)
                         if kernel_id is None:
                             continue
+                        if evdata['Action']=='oom':
```
Comment (achimnol) [inline src/ai/backend/agent/docker/agent.py:1231]: Please keep PEP-8 code style.

File: src/ai/backend/agent/docker/agent.py [hunk 2]
```diff
@@ -1228,6 +1228,9 @@ async def handle_action_die(kernel_id: KernelId, evdata: Mapping[str, Any]) -> N
...
+                            kernel_obj = self.kernel_registry.get(kernel_id)
+                            kernel_obj.termination_reason = 'outofmemory'
```
Comment (achimnol) [inline src/ai/backend/agent/docker/agent.py:1233]: The reason uses a "kebab-case" style. Let's change it to `'out-of-memory'`.

**Review by achimnol — CHANGES_REQUESTED** [review]

**Comment by achimnol** [comment]
I have questions:
* Do both events ("oom" and "die") happen for a container killed by oom?
* If so, is the order of "oom" and "die" always consistent?

**Comment by youngjun0627** [comment]
Yes, OOM killer decides to kill the container when catching out-of-memory. So, the order is 'create -> start -> oom -> kill -> die -> stop -> destroy'

File: src/ai/backend/agent/docker/agent.py (hunk 2 above)
Comment (youngjun0627) [reply src/ai/backend/agent/docker/agent.py]: I have finished to change it!

**Comment by achimnol** [comment]
Thanks for the answers.
I'm just worried about whether the termination reason is overwritten or not.
Have you confirmed from the client by checking the `status_info` field of sessions?

**Comment by youngjun0627** [comment]
I have a question.
- if the container has many process,  OOM killer remove one or some processes.  Then, will we decided to terminate the container? If not, the container maintains status of 'RUNNING' and stays until client requests to destroy it.

The implemented code updates column of status-info on 'kernels' table. But, because the container is terminated by client-destroy-request, status-info of 'user-requested' is displayed to client.

**Comment by achimnol** [comment]
Ah... then, can the OOM event occur *multiple* times for a single container? Hm.........

**Comment by youngjun0627** [comment]
Yes.
For example, if three processes(task) run in a container and OOM event occurs **(one)**, OOM killer chooses to terminate one or more of them according to priority. 
After one process is removed by OOM killer, if another process is out-of-memory, OOM event occurs **(two)** and OOM killer does it.
Finally, if the other process is not out-of-memory, the container has one-process is running . And if not **(three)**, the container does not have a process, but it is running.
**OOM killer removes process(task), not container.** 
However, I think all the task in kernel made by client is important. So, what do you think about terminating the container immediately when it finds 'oom'?

**Comment by achimnol** [comment]
Directly killing the container upon every OOM events may be too harsh for some deep learning workloads such as PyTorch, as @inureyes says on a discussion with him.
We might need to add a new feature such as setting a shell command for each session to be executed when OOM occurs, so that we could specify user-defined recovery actions. Just let's hold the issue until we make design decision.

**Comment by achimnol** [comment]
Please resolve the conflicts, checking the functionality of recent main branch updates such as #299, #300, #301.

**Comment by youngjun0627** [comment]
- Idea
when OOM event occurs, container is paused(of course, status-info is 'out-of-memory').
The user checks the sesision status and unpauses or restarts the container after the problem is solved.
So, I suggest we add a command option(unpause) and a container status('PAUSED')

**Comment by youngjun0627** [comment]
image

I add 'paused' and 'unpaused' status for client to recover it when 'OOM' occurs. For this idea, we need to add PausedEvent and the handler of this. I wonder your idea about this issue or feedback from my solution.

**Comment by achimnol** [comment]
Migrated to lablup/backend.ai#629.

**Pull request closed without merging**